src/back/msl/writer.rs

```diff
@@ -1061,25 +1063,37 @@ impl<W: Write> Writer<W> {
             // has to be a named expression
             crate::Expression::Call(_) => unreachable!(),
             crate::Expression::ArrayLength(expr) => match *context.resolve_type(expr) {
-                crate::TypeInner::Array {
-                    size: crate::ArraySize::Constant(const_handle),
-                    ..
-                } => {
-                    let coco = ConstantContext {
-                        handle: const_handle,
-                        arena: &context.module.constants,
-                        names: &self.names,
-                        first_time: false,
-                    };
-                    write!(self.out, "{}", coco)?;
-                }
-                crate::TypeInner::Array { .. } => {
-                    return Err(Error::FeatureNotImplemented(
-                        "dynamic array size".to_string(),
-                    ))
-                }
+                crate::TypeInner::Pointer { base, .. } => match context.module.types[base].inner {
+                    crate::TypeInner::Array { stride, .. } => {
+                        let handle = context.info[expr].assignable_global.unwrap();
```

Comment:
Not sure if exposing the assignable global (in public!) is justified for this case.

Perhaps, we could go another way here. Instead of asking about the type of `expr` and traversing the types from there, we could first match the expression itself, then reach to `Expression::GlobalVariable`, and that gives us both the type handle (to match for a struct) and the lookup key into `runtime_sized_buffers`.

How do you feel about trying this out? This would keep `assignable_global` private, and it seems conceptually more straightforward.